### PR TITLE
Unit test deduper

### PR DESCRIPTION
**Pull request by mollerhoj**
Depends on https://github.com/centre-for-humanities-computing/danish-foundation-models/pull/46

This PR adds unittests and general improvements to the deduper.

# Code improvements (subjective)
- Extracted a few methods to make tests and comprehension easier

# Features
- `silent` parameter to not print progress (useful for testing)
- `normalization_func` parameter to give user defined callback as document normalization. The old normalization has been set as the default function

# Bugfixes
- Shingles no longer forgets the last token ("abc" of char ngram size 1 becomes "a,b,c" not just "a,b")
- For documents shorter than the shingle size, we now return a single shingle ("Hej med dig" of char_ngram_size 13 becomes ["Hej med dig"] not just [].) This avoid deduplicating short but completely different documents (They would all return the same empty fingerprint).

### CODE REVIEW

**Comment by mollerhoj** [comment]
As mentioned here: https://github.com/centre-for-humanities-computing/danish-foundation-models/discussions/45 it might makes sense to replace (or extend) the split_method to use https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.HashingVectorizer.html instead.

**Review by saattrupdan — CHANGES_REQUESTED** [review]
Nice one, good call with the option to change normalisation! I like the tests as well, especially the "miss-tests" when varying the minhash byte size, good idea 😊

File: tests/cleaning/deduper_test.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+import pytest
+import tempfile
+from pathlib import Path
+import json
+import re
+
+
+class TestDeduper:
+    def deduper(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (mollerhoj) [reply tests/cleaning/deduper_test.py:11]: The efficiency gains here will be neglectable, but I appreciate if it's more idiomatic. I wouldn't wanna sacrifice the conciseness of individual unit tests though.

I'm not sure I understand though: `deduper.ngram_size = 13` is an assignment *after* initialization?

Feel free to add commits to this PR if you wanna demonstrate.

File: dfm/cleaning/deduper.py [hunk 2]
```diff
@@ -20,6 +20,20 @@
 from unicodedata import normalize
 import re
 from tqdm.auto import tqdm
+from collections.abc import Callable
```
Comment (saattrupdan) [inline dfm/cleaning/deduper.py:23]: This should be imported from `typing` instead (causing an error on my side):

**Review by saattrupdan — CHANGES_REQUESTED** [review]
Let's get this PR sorted. There's a single import bug, and you need to insert (blank) `__init__.py` scripts in the subfolders of the `tests` folders, as otherwise `pytest` doesn't know how to import scripts. Otherwise, I think it's fine 🙂
